In [5]:
pip install jovian


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 68 kB 4.5 MB/s 
  Created wheel for uuid: filename=uuid-1.30-py3-none-any.whl size=6503 sha256=59078af62acef26ee0b098ceaba89de9a5d0ab9930d18d7e2a945db9fb9f0865
  Stored in directory: /root/.cache/pip/wheels/2a/ea/87/dd57f1ecb4f0752f3e1dbf958ebf8b36d920d190425bcdc24d
Successfully built uuid


In [6]:
import jovian


# **Linear Search and Binary Search**

# **Creating Test Cases**

In [ ]:
def locate_card(cards,query):
  pass

In [ ]:
test = {
    'input': {
        'cards': [13,11,10,7,4,2,1,0],
        'query': 7
        },
        'output':3
    }


In [ ]:
locate_card(**test['input']) == test['output']

False

Adding more test cases

We created tests list because you cannot append in dictionary


In [ ]:
tests=[]

In [ ]:
tests.append({
    'input':{
        'cards':[6],
        'query' : 6
        },
        'output':0
    })

In [ ]:
tests.append(test)

In [ ]:
tests.append({
    'input': {
        'cards': [8, 8, 6, 6, 6, 6, 6, 3, 2, 2, 2, 0, 0, 0],
        'query': 3
    },
    'output': 7
})
tests.append({
    'input': {
        'cards': [],
        'query': 7
    },
    'output': -1
})

In [ ]:
tests


[{'input': {'cards': [6], 'query': 6}, 'output': 0},
 {'input': {'cards': [13, 11, 10, 7, 4, 2, 1, 0], 'query': 7}, 'output': 3},
 {'input': {'cards': [8, 8, 6, 6, 6, 6, 6, 3, 2, 2, 2, 0, 0, 0], 'query': 3},
  'output': 7},
 {'input': {'cards': [], 'query': 7}, 'output': -1}]

**Brute force - Linear Search**

In [ ]:
def locate_card(cards,query):
  position = 0
  while True:
    if cards[position]==query:
      return position
    position+=1
    if position == len(cards):
      return -1

In [ ]:
result= locate_card(test['input']['cards'],test['input']['query'])
result

3

In [ ]:
!pip install jovian --upgrade --quiet

In [7]:
from jovian.pythondsa import evaluate_test_case

In [ ]:
evaluate_test_case(locate_card,test)


Input:
{'cards': [13, 11, 10, 7, 4, 2, 1, 0], 'query': 7}

Expected Output:
3


Actual Output:
3

Execution Time:
0.005 ms

Test Result:
PASSED



(3, True, 0.005)

In [8]:
from jovian.pythondsa import evaluate_test_cases

In [ ]:
evaluate_test_cases(locate_card,tests)

debugging

In [ ]:
def locate_card(cards,query):
  position = 0
  print('cards:',cards)
  print('query:',query)
  while True:
    print('position:',position)
    if cards[position]==query:
      return position
    position+=1
    if position == len(cards):
      return -1

In [ ]:
cards3 = tests[3]['input']['cards']
query3 = tests[3]['input']['query']
locate_card(cards3,query3)

# **Correct way for Linear Search**

In [ ]:
def locate_card(cards,query):
  position = 0
  print('cards:',cards)
  print('query:',query)
  while position<len(cards):
    print('position:',position)
    if cards[position]==query:
      return position
    position+=1
    
  return -1

In [ ]:
cards3 = tests[3]['input']['cards']
query3 = tests[3]['input']['query']
locate_card(cards3,query3)

In [ ]:
evaluate_test_cases(locate_card,tests)


TEST CASE #0
cards: [6]
query: 6
position: 0

Input:
{'cards': [6], 'query': 6}

Expected Output:
0


Actual Output:
0

Execution Time:
0.07 ms

Test Result:
PASSED


TEST CASE #1
cards: [13, 11, 10, 7, 4, 2, 1, 0]
query: 7
position: 0
position: 1
position: 2
position: 3

Input:
{'cards': [13, 11, 10, 7, 4, 2, 1, 0], 'query': 7}

Expected Output:
3


Actual Output:
3

Execution Time:
0.129 ms

Test Result:
PASSED


TEST CASE #2
cards: [8, 8, 6, 6, 6, 6, 6, 3, 2, 2, 2, 0, 0, 0]
query: 3
position: 0
position: 1
position: 2
position: 3
position: 4
position: 5
position: 6
position: 7

Input:
{'cards': [8, 8, 6, 6, 6, 6, 6, 3, 2, 2, 2, 0, 0, 0], 'query': 3}

Expected Output:
7


Actual Output:
7

Execution Time:
0.206 ms

Test Result:
PASSED


TEST CASE #3
cards: []
query: 7

Input:
{'cards': [], 'query': 7}

Expected Output:
-1


Actual Output:
-1

Execution Time:
0.041 ms

Test Result:
PASSED


SUMMARY

TOTAL: 4, PASSED: 4, FAILED: 0


[(0, True, 0.07), (3, True, 0.129), (7, True, 0.206), (-1, True, 0.041)]

In [ ]:
jovian.commit(project='python-binary-search', environment=None)

[jovian] Detected Colab notebook...


[jovian] Error: jovian.commit doesn't work on Colab unless the notebook was created and executed from Jovian.
Make sure to run the first code cell at the top after executing from Jovian.
Alternatively, you can download this notebook and upload it manually to Jovian.
Learn more: https://jovian.ai/docs/user-guide/run.html#run-on-colab


# **Binary Search**

In [ ]:
def locate_card(cards,query):
  lo,hi=0,len(cards)-1
  while lo <= hi:
    mid = (lo + hi)//2
    mid_num = cards[mid]
    print('lo:',lo,", hi :",hi,", mid:",mid,",mid_num:",mid_num)
    if mid_num == query:
      return mid
    elif mid_num < query:
      hi = mid -1
    elif mid_num > query:
      lo = mid + 1
  return -1

In [ ]:
evaluate_test_cases(locate_card, tests)



TEST CASE #0
lo: 0 , hi : 0 , mid: 0 ,mid_num: 6

Input:
{'cards': [6], 'query': 6}

Expected Output:
0


Actual Output:
0

Execution Time:
0.156 ms

Test Result:
PASSED


TEST CASE #1
lo: 0 , hi : 7 , mid: 3 ,mid_num: 7

Input:
{'cards': [13, 11, 10, 7, 4, 2, 1, 0], 'query': 7}

Expected Output:
3


Actual Output:
3

Execution Time:
0.142 ms

Test Result:
PASSED


TEST CASE #2
lo: 0 , hi : 13 , mid: 6 ,mid_num: 6
lo: 7 , hi : 13 , mid: 10 ,mid_num: 2
lo: 7 , hi : 9 , mid: 8 ,mid_num: 2
lo: 7 , hi : 7 , mid: 7 ,mid_num: 3

Input:
{'cards': [8, 8, 6, 6, 6, 6, 6, 3, 2, 2, 2, 0, 0, 0], 'query': 3}

Expected Output:
7


Actual Output:
7

Execution Time:
0.928 ms

Test Result:
PASSED


TEST CASE #3

Input:
{'cards': [], 'query': 7}

Expected Output:
-1


Actual Output:
-1

Execution Time:
0.003 ms

Test Result:
PASSED


SUMMARY

TOTAL: 4, PASSED: 4, FAILED: 0


[(0, True, 0.156), (3, True, 0.142), (7, True, 0.928), (-1, True, 0.003)]

In [ ]:
tests.append({
    'input': {
        'cards':[8, 8, 6, 6, 6, 6, 6, 6, 3, 2, 2, 2, 0, 0, 0],
        'query':6
    },
    'output':2
})

In [ ]:
evaluate_test_cases(locate_card,tests)


TEST CASE #0
lo: 0 , hi : 0 , mid: 0 ,mid_num: 6

Input:
{'cards': [6], 'query': 6}

Expected Output:
0


Actual Output:
0

Execution Time:
0.154 ms

Test Result:
PASSED


TEST CASE #1
lo: 0 , hi : 7 , mid: 3 ,mid_num: 7

Input:
{'cards': [13, 11, 10, 7, 4, 2, 1, 0], 'query': 7}

Expected Output:
3


Actual Output:
3

Execution Time:
0.148 ms

Test Result:
PASSED


TEST CASE #2
lo: 0 , hi : 13 , mid: 6 ,mid_num: 6
lo: 7 , hi : 13 , mid: 10 ,mid_num: 2
lo: 7 , hi : 9 , mid: 8 ,mid_num: 2
lo: 7 , hi : 7 , mid: 7 ,mid_num: 3

Input:
{'cards': [8, 8, 6, 6, 6, 6, 6, 3, 2, 2, 2, 0, 0, 0], 'query': 3}

Expected Output:
7


Actual Output:
7

Execution Time:
0.692 ms

Test Result:
PASSED


TEST CASE #3

Input:
{'cards': [], 'query': 7}

Expected Output:
-1


Actual Output:
-1

Execution Time:
0.003 ms

Test Result:
PASSED


TEST CASE #4
lo: 0 , hi : 14 , mid: 7 ,mid_num: 6

Input:
{'cards': [8, 8, 6, 6, 6, 6, 6, 6, 3, 2, 2, 2, 0, 0, 0], 'query': 6}

Expected Output:
2


Actual Output:
7

Execu

[(0, True, 0.154),
 (3, True, 0.148),
 (7, True, 0.692),
 (-1, True, 0.003),
 (7, False, 0.18)]

In [ ]:
cards4 = tests[4]['input']['cards']
query4 = tests[4]['input']['query']


In [ ]:
cards4[7]

6

So we did locate a '6' in the array but it seems like there was an early occurence as well but since we were just using mid to find query we stopped at 7th position only.

Correct way for Binary Search

In [ ]:
def test_location(cards, query, mid):
    mid_number = cards[mid]
    #print("mid:", mid, ", mid_number:", mid_number)
    if mid_number == query:
        if mid-1 >= 0 and cards[mid-1] == query:
            return 'left'
        else:
            return 'found'
    elif mid_number < query:
        return 'left'
    else:
        return 'right'

def locate_card(cards, query):
    lo, hi = 0, len(cards) - 1
    
    while lo <= hi:
        #print("lo:", lo, ", hi:", hi)
        mid = (lo + hi) // 2
        result = test_location(cards, query, mid)
        
        if result == 'found':
            return mid
        elif result == 'left':
            hi = mid - 1
        elif result == 'right':
            lo = mid + 1
    return -1

In [ ]:
evaluate_test_cases(locate_card,tests)

# **Linear Search vs Binary Search**

Linear Search

In [1]:
def locate_card_linear(cards,query):
  position = 0
  while position <= len(cards):
    if cards[position] == query:
      return position
    position+=1
  return -1


In [3]:
large_test={
    'input' : {
        'cards' : list(range(10000000,0,-1)),
        'query' : 2
    },
    'output' : 9999998
}

In [12]:
evaluate_test_case(locate_card_linear,large_test)


Input:
{'cards': [10000000, 9999999, 9999998, 9999997, 9999996, 9999995, 9999994, 9999993, 9999992, 9999991...

Expected Output:
9999998


Actual Output:
9999998

Execution Time:
1927.496 ms

Test Result:
PASSED



(9999998, True, 1927.496)

Binary Search

In [10]:
def test_location_binary(cards,query,mid):
  mid_number = cards[mid]
  if cards[mid] == query:
    if mid-1 >= 0 and cards[mid-1] == query:
      return 'left'
    else:
      return 'found'
  elif cards[mid] < query:
    return 'left'
  else:
    return 'right'

def locate_card_binary(cards,query):
  lo,hi = 0,len(cards)-1
  while lo <= hi:
    mid = (lo+hi)//2
    result = test_location_binary(cards,query,mid)

    if result == 'found':
      return mid
    elif result == 'left':
      hi = mid -1
    elif result == 'right':
      lo = mid+1
  return -1


In [11]:
evaluate_test_case(locate_card_binary,large_test)


Input:
{'cards': [10000000, 9999999, 9999998, 9999997, 9999996, 9999995, 9999994, 9999993, 9999992, 9999991...

Expected Output:
9999998


Actual Output:
9999998

Execution Time:
0.033 ms

Test Result:
PASSED



(9999998, True, 0.033)

# **Execution time for Binary is 0.033 ms while for linear is was 1927.496 ms**